### DISCUSSION TITLE
A few questions on the MetaLayer

### DISCUSSION THREAD

**Discussion by alzaia** [pyg-team/pytorch_geometric#6021] [Q&A] [answered]
Hello, I was able to adapt the `MetaLayer` to my actual needs but still have a few practical questions to make sure that I'm making the most of it. 
1. Currently I did not implement the `GlobalModel` and I'm sticking to the `NodeModel` and `EdgeModel` (basically doing the steps 1 to 3 of the Battaglia paper algorithm 1: update the edges, aggregate edge attributes per node, and then update the nodes). I followed the PyG example of the `MetaLayer` closely. One `MetaLayer` pass performs one update for both edges and nodes. However, I'm stacking multiple `MetaLayer` layers in my case to get a more expressive model (no weight sharing: different `EdgeModel` and `NodeModel` modules are used for each block). Is this the intended use of the `MetaLayer`?
2. On the `EdgeModel`: are there other options for the NN here? I want ideally something more expressive than a MLP. Or at least stacking more of them here?
3. On the `NodeModel`: is it doable here to use another existing node update layer (e.g. GCN, GAT) instead of the MLP examples provided in the PyG doc? If yes, what would I need to change from the doc version of the `NodeModel`?

Thanks!

**Comment by wsad1** [comment, marked as answer]
1. I think you are using it as intended. The whole idea of `MetaLayer` is to plug in blocks you like, depending on your use case.
2. You could stack mlps and also maybe take difference/dot product between node features to make it more expressive in practice.
3. Yes you could replace it with any model given that model has the following forward interface `forward(self, x, edge_index, edge_attr, u, batch)`. For  example, in `GATConv` the interface is `forward(x, edge_index, edge_attr, size, return_attention_wieghts)` which should work as `NodeModel`, given `u` is `None`. If its not `None` `u` will get passed as `size` which might cause issues, but I am not sure of that.

**Reply by alzaia** [reply]
Thanks a lot @wsad1!
